### PR TITLE
Fencing a failed primary

### PR DESCRIPTION
**Pull request by davissp14**
This PR aims to ensure that a booting primary that has diverged from the cluster is properly isolated.

Note: There's still some cleanup to do, but the core logic is there. 

## What this doesn't solve
Split-brains as a result of a network partition.  

## How do we verify the real primary?
We start out evaluating the cluster state by checking each registered standby for connectivity and asking who their primary is.

The "clusters state" is represented across a few different dimensions:

**Total members**    
Number of registered members, including the primary.

**Total active members**
Number of members that are responsive.  This includes the primary we are evaluating, so this will never be less than one.

**Total inactive members**
Number of registered members that are non-responsive. 

**Conflict map**
The conflict map is a `map[string]int` that tracks conflicting primary's queried from our standbys and the number of occurrences a given primary was referenced.

As an example, say we have a 3 member cluster and both of the standby's indicate their registered primary does not match.  This will be recorded as: 
```
map[string]int{
  "fdaa:0:2e26:a7b:8c31:bf37:488c:2": 2
}
```

The real primary is resolvable so long as the majority of members can agree on who it is.  Quorum being defined as `total_members / 2 + 1`.

**There is one exception to note here.  When self meets quorum, we will still fence ourself in the event a conflict is found.  This is to protect against a possible race condition if we are coming up during an active failover.** 

Tests can be found here: https://github.com/fly-apps/postgres-flex/pull/49/files#diff-3d71960ff7855f775cb257a74643d67d2636b354c9d485d10c2ded2426a7f362

## What if the real primary can't be resolved or doesn't match the booting primary?

In both of these instances the primary member will be fenced. 

**If the real primary is resolvable** 
The cluster will be made read-only, the PGBouncer will be reconfigured to target the "real" primary and the ip address is written to a `zombie.lock` file.  The PGBouncer reconfiguration will ensure that any connections hitting this member will be routed to the real primary in order to minimize interruptions.  Once, completed there will be panic to force a full member restart. When the member is restarted, we will read the ip address from the `zombie.lock` file and use that to attempt to rejoin the cluster we diverged from.  If we are successful, the `zombie.lock` is cleared and we will boot as a standby.    

**Note: We will not attempt to rejoin a cluster if the resolved primary resides in a region that differs from the `PRIMARY_REGION` environment variable set on self.  The `PRIMARY_REGION` will need to be updated before a rejoin will be attempted.**

**If the real primary is NOT resolvable**
The cluster will be made read-only, PGBouncer will remain disabled and a `zombie.lock` file will be created without a value.  When the member reboots, we will read the `zombie.lock` file and see that it's empty.  This indicates that we've entered a failure mode that can't be recovered automatically.  This could be an issue where previously deleted members were not properly unregistered, or the primary's state has diverged to a point where its registered members have been cycled out.   

How to address these situations will be a part of future documentation.

### CODE REVIEW

File: pkg/flypg/node.go [hunk 1]
```diff
@@ -234,10 +270,98 @@ func (n *Node) PostInit(ctx context.Context) error {
...
+					// Create a zombie.lock file containing the resolved primary.
+					// This will be an empty string if we are unable to resolve the real primary.
+					if err := writeZombieLock(primary); err != nil {
+						return fmt.Errorf("failed to set zombie lock: %s", err)
+					}
+
+					fmt.Println("Setting all existing tables to read-only")
+					if err := admin.SetReadOnly(ctx, conn); err != nil {
+						return fmt.Errorf("failed to set read-only: %s", err)
+					}
```
Comment (DAlperin) [inline pkg/flypg/node.go:345]: I think we shouldn't return if we can't reconfigure pgbouncer here, since we should still set existing tables to readonly

File: pkg/flypg/node.go [hunk 2]
```diff
@@ -234,10 +270,98 @@ func (n *Node) PostInit(ctx context.Context) error {
...
+			totalConflicts := 0
+			conflictMap := map[string]int{}
+
+			for _, standby := range standbys {
+				mConn, err := repmgr.NewRemoteConnection(ctx, standby.Hostname)
+				if err != nil {
+					fmt.Printf("failed to connect to %s", standby.Hostname)
+					totalInactive++
+					continue
+				}
```
Comment (benbjohnson) [inline pkg/flypg/node.go:296]: Should these `mConn` connections be closed? It doesn't look like they're used after passing to `PrimaryMember()`.

File: pkg/flypg/node.go [hunk 3]
```diff
@@ -234,10 +270,98 @@ func (n *Node) PostInit(ctx context.Context) error {
...
+					fmt.Println("Setting all existing tables to read-only")
+					if err := admin.SetReadOnly(ctx, conn); err != nil {
+						return fmt.Errorf("failed to set read-only: %s", err)
+					}
+
+					return fmt.Errorf("zombie primary detected. Use `fly machines restart <machine-id>` to rejoin the cluster or consider removing this node")
+				}
+
+				return fmt.Errorf("failed to run zombie diagnosis: %s", err)
+			}
```
Comment (benbjohnson) [inline pkg/flypg/node.go]: You can reduce nesting a bit by moving the `errors.Is()` outside the outer `if`:

```go
if errors.Is(err, ErrZombieDiscovered) {
	...
} else if err != nil {
	...
}
```

File: pkg/flypg/node.go [hunk 4]
```diff
@@ -234,10 +270,98 @@ func (n *Node) PostInit(ctx context.Context) error {
 		if member != nil && member.Role != "" {
 			role = member.Role
 		}
+		fmt.Printf("My current role is: %s\n", role)
 
 		switch role {
 		case PrimaryRoleName:
-			// TODO - This is where we need to fence the primary if it's no longer valid.
+			standbys, err := repmgr.StandbyMembers(ctx, conn)
```
Comment (DAlperin) [inline pkg/flypg/node.go:277]: still building my mental model here but shouldn't we query all members in the case that the local repmgr thinks the current "real" primary is now a stanby or vice-versa?

File: pkg/flypg/node.go [hunk 5]
```diff
@@ -322,6 +446,11 @@ func (n *Node) configure(store *state.Store) error {
 		fmt.Println(err.Error())
 	}
 
+	// Clear target and wait for primary resolution
+	if err := n.PGBouncer.ConfigurePrimary(ctx, "", false); err != nil {
+		fmt.Println(err.Error())
```
Comment (benbjohnson) [inline pkg/flypg/node.go:450]: Does this not need to be bubbled up?

File: pkg/flypg/zombie.go [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+	return true
+}
+
+func writeZombieLock(hostname string) error {
+	if err := ioutil.WriteFile("/data/zombie.lock", []byte(hostname), 0644); err != nil {
+		return err
+	}
+
+	return nil
+}
```
Comment (benbjohnson) [inline pkg/flypg/zombie.go:23]: Minor nit. `ioutil` is deprecated. You can use `io.WriteFile()` instead.

File: pkg/flypg/zombie_test.go [hunk 7]
```diff
@@ -0,0 +1,243 @@
+package flypg
+
+import (
+	"errors"
+	"fmt"
+	"testing"
+)
+
+func TestZombieDiagnosis(t *testing.T) {
```
Comment (DAlperin) [inline pkg/flypg/zombie_test.go:9]: this is excellent. this makes it easier for me to continue to threaten to make real integration tests :)

File: pkg/flypg/zombie_test.go [hunk 8]
```diff
@@ -0,0 +1,243 @@
...
+				if !c.ExpectedZombie {
+					fmt.Printf("Hostname: %s, Total members: %d, Total active: %d, Total inactive: %d, Conflicts: %+v\n",
+						c.MyHostname,
+						c.TotalMembers,
+						c.TotalActive,
+						c.TotalInactive,
+						c.ConflictMap,
+					)
+					t.Logf("test case: %d failed. Wasn't expecting to be a Zombie", i)
+					t.Fail()
```
Comment (benbjohnson) [inline pkg/flypg/zombie_test.go:218]: These can be combined into a `t.Fatalf()`

File: pkg/flypg/zombie_test.go [hunk 9]
```diff
@@ -0,0 +1,243 @@
...
+				ConflictMap: map[string]int{
+					"secret-host": 3,
+				},
+				ExpectedHostname: "secret-host",
+				ExpectedZombie:   true,
+			},
+		},
+	}
+
+	for i, c := range tests.Cases {
```
Comment (benbjohnson) [inline pkg/flypg/zombie_test.go:204]: I'm not personally a fan of table tests for anything but really simple tests. I find they get really complicated as you start adding edge cases. I'd probably rework these into subtests using `t.Run()`.

Although, if you want to keep the table tests, you can still wrap each of these in a `t.Run()` and then you can filter on them with `go test -run=XXX`

File: pkg/flypg/node.go [hunk 10]
```diff
@@ -322,6 +445,11 @@ func (n *Node) configure(store *state.Store) error {
 		fmt.Println(err.Error())
 	}
```
Comment (DAlperin) [inline pkg/flypg/node.go:446]: here too I suppose

File: pkg/flypg/node.go (hunk 4 above)
Comment (davissp14) [reply pkg/flypg/node.go:277]: Within this context, the local repmgr is identifying itself as the primary.

File: pkg/flypg/node.go (hunk 1 above)
Comment (davissp14) [reply pkg/flypg/node.go]: Yeah, i'll need to re-evaluate the returns a bit more.  Good call.

File: pkg/flypg/node.go (hunk 4 above)
Comment (DAlperin) [reply pkg/flypg/node.go:277]: ah right, thanks

File: pkg/flypg/node.go (hunk 2 above)
Comment (davissp14) [reply pkg/flypg/node.go:296]: Yes they should!

File: pkg/flypg/zombie.go (hunk 6 above)
Comment (davissp14) [reply pkg/flypg/zombie.go:23]: Oh, right.  Are you talking about `os.WriteFile` ?

**Comment by davissp14** [comment]
One thing we may need to consider here is region. 🤔 

Say we have a 3 member cluster and two of the standbys reside outside of the primary region.  Should the primary still require quorum, or should we treat this like a single member cluster.

**Comment by davissp14** [comment]
This seems to work well overall, however, I was able to discover a race condition that can happen in the event the failed primary comes back up in the middle of a failover.  

Example: 3 member setup

Node A: Primary
Node B: Standby
Node C: Standby

Node A - Fails
Node B  - Is elected Primary
Node A - Comes back to life before Node C is reconfigured. 
Node A - Is able to meet quorum and continues being primary.
Node C - Is reconfigured to follow Node B.

This being the case, I think we need to take a slightly more conservative approach and fence the primary in the event any primary conflicts exist, even when quorum is met.  This

**Review by DAlperin — APPROVED** [review]
